ui/src/source.ts

```diff
@@ -379,29 +297,71 @@ export const revisionQueryEq = (
   }
 };
 
-export const readme = (
-  projectId: string,
-  peerId: string,
-  revision: RevisionQuery
-): Readable<remote.Data<Readme | null>> => {
-  const readme = remote.createStore<Readme | null>();
-
-  remote
-    .chain(objectStore.readable, readme)
-    .then((object: SourceObject) => {
-      if (object.info.objectType === ObjectType.Tree) {
-        const path = findReadme(object as Tree);
-
-        if (path) {
-          return blob(projectId, peerId, revision, path, false);
-        }
-      }
-
-      return null;
-    })
-    .then(blob => (blob && !blob.binary ? blob : null))
-    .then(readme.success)
-    .catch(readme.error);
+export const fetchReadme = (
+  projectUrn: Urn,
+  peerId: PeerId,
+  revision: Revision,
+  tree: Tree
+): Promise<Readme | null> => {
+  const path = findReadme(tree);
+
+  return new Promise((resolve, _reject) => {
+    if (!path) {
+      return resolve(null);
+    }
 
-  return readme.readable;
+    fetchBlob(projectUrn, peerId, path, revision, false)
+      .then(blob => (blob && !blob.binary ? blob : null))
+      .then(resolve)
+      .catch(() => resolve(null));
```

Comment:
We should be more discriminate here: Only if the readme does not exist should we return `null`. Otherwise we should notify the user of an error.

Reply:
Indeed!